### PR TITLE
add a switch to disable all output docs in fastapi

### PR DESCRIPTION
**Pull request by Zheaoli**
add a switch to disable output docs in the production environment. They can use the system environment variable or use  `disable_all_output_docs` variable  to disable output docs in an explicit way

### CODE REVIEW

File: fastapi/applications.py [hunk 1]
```diff
@@ -85,6 +87,7 @@ def __init__(
         if self.docs_url or self.redoc_url:
             assert self.openapi_url, "The openapi_url is required for the docs"
         self.openapi_schema: Optional[Dict[str, Any]] = None
+        self.__disable_all_outpu_docs = disable_all_output_docs
```
[the remark this answers is not in the document]
Comment (Zheaoli) [reply fastapi/applications.py:90]: nice touch, I'll fix it

**Comment by phy25** [comment]
So fastapi never had this `FASTAPT_ENVIRONMENT` before, and I doubt if production equals to no openapi.

Also I am not sure this quick path is worth implementing. You can always implement this in your own application logic by [passing in `openapi_url`, `docs_url`, `redoc_url`](https://fastapi.tiangolo.com/tutorial/metadata/#openapi-url). I think it's encouraged to use Pydantic's `BaseConfig` to do the envvar reading.

**Comment by Zheaoli** [comment]
@phy25  I think it's worth to implement it. 

Because people want to use the same code in local environment and production environment

But you are right, the production doesn't mean that we don't need openapi doc

So I can split it into four  environment variables

1. FASTAPI_DISABLE_ALL_DOCS

2. FASTAPI_DISABLE_OPENAPI_DOCS

3. FASTAPI_DISABLE_SWAGGER_DOCS

4. FASTAPI_DISABLE_REDOCS

I think it's more convenient for people that they want to control the docs visibility simply without any extra code

**Comment by phy25** [comment]
Reference on `pydantic.BaseSettings`: https://fastapi.tiangolo.com/advanced/settings/#pydantic-settings

**Comment by Zheaoli** [comment]
@phy25 Actually I didn't get your point about pedantic?

**Comment by phy25** [comment]
I just think fastapi prefers using pydantic for config - when you read envvars you may eventually want to read .env files. So probably in the first place I don't think this feature is worth implementing.

Anyway it's just my own opinion.

**Comment by Zheaoli** [comment]
I see, maybe you mean that we can use pydantic-settings to instead of processing the environment manually?

**Comment by Zheaoli** [comment]
> I just think fastapi prefers using pydantic for config - when you read envvars you may eventually want to read .env files.

@phy25  that's a good way, I will update my code for this PR 

> So probably in the first place I don't think this feature is worth implementing.

Actually, I have already discuss this feature with my friends who are FastAPI's users too, they think they need a simple way to disable the docs output in the production environment without any extra code.

**Comment by phy25** [comment]
Again, you may have other production-related config change besides this docs settings, and this can be easily implemented with custom app logic. I still don't think this is worth implementing in the core - my humble opinion.

**Comment by Zheaoli** [comment]
> you may have other production-related config change besides this docs settings, and this can be easily implemented with custom app logic.

@phy25 this is making sense.

But in my humble personal opinion, the openapi doc, the swagger doc is one of the core features in this framework, and it's deeply about the security problem for the doc in production. That's means people need to care about the accessibility in the sever with FastAPI by disableing the doc or using other ways.

Imagine this, I have a couple of servers made by FastAPI. I want to disable the docs in the production environment( most of the people will do this), this means I will get some duplicate code across the codebase, just like this 

```python

class Settings(BaseSettings):
    environment: str = "DEBUG"


setting = Settings()


def create_app():
    if setting.environment == "DEBUG":
        app = FastAPI(openapi_url=None, docs_url=None, redoc_url=None)
    else:
        app = FastAPI
    return app
```

I think it's not a good way to people. Most of the people want to get a simple way to control the security-sensitive features like the doc. I guess this is not a technical problem, but it's a user experience problem.

Thanks for your review, hope to get more ideas from you. Your idea is very useful and important for me! :)

**Comment by Dustyposa** [comment]
Awesome, at least, this PR is very useful for me!
Because, in production env, i must disable it that exposed my api maybe cause some safe problem.
For me, to avoiding this problem, I use this way:
```python
DEBUG = config("DEBUG", cast=bool, default=False)
if DEBUG:
    DOCS_URL = "/docs"
else:
    DOCS_URL = ""

...
app = Fastapi(docs_url=DOCS_URL)

```
Code is small, but usually need do this for me, also for other user?(in my opinion)
I think it's not very readable, other people need to check the `DOCS_URL's` value.
So, I think do a switch to do this can be more readable and use easy.
Like `app = Fastapi(disable_doc=True)` .    
:)

**Comment by Zheaoli** [comment]
@tiangolo ping ~

**Comment by heitorPB** [comment]
This PR is also very useful for me.

The only thing odd is the double negation to enable the docs. For example:
`FASTAPI_DISABLE_REDOCS = False` enables ReDoc. But I would prefer `FASTAPI_ENABLE_REDOCS = True` to enable it and `FASTAPI_ENABLE_REDOCS = False` to disable it.

Setting an `ENABLE_X` variable to `True` to enable feature X is more readable than setting `DISABLE_X` to `False` to enable it.

What do you think?

**Comment by Zheaoli** [comment]
@heitorPB this is making sense, I will update the desgin

**Comment by Zheaoli** [comment]
@tiangolo ping~

**Comment by tiangolo** [comment]
Hey @Zheaoli , thanks for your interest and effort! :nerd_face: 

And thanks everyone for the discussion here. :coffee: 

---

I wouldn't think that disabling docs for production would necessarily be a good practice, I think any security stuff should be implemented in the backend by adding the right tokens, encryption, even scopes if necessary, instead of by hiding the docs.

But anyway, I can see people could like to do it. Nevertheless, I would prefer not to add predefined FastAPI-specific env vars to disable or enable docs as I think the specific needs for something like this would vary quite a lot in a case by case basis. I think in many scenarios it would make sense to even override docs completely and provide them in a custom path operation, e.g. to add Cookie-based authentication so that your own developers can still access the docs.

---

On the other side, this made me realize that disabling docs requires disabling the 3 parameters, so I added support for disabling OpenAPI and the docs by just disabling the `openapi_url`.

I also added docs and tests for it: https://fastapi.tiangolo.com/advanced/conditional-openapi/

So you can implement it with something as simple as:

```Python
from fastapi import FastAPI
from pydantic import BaseSettings


class Settings(BaseSettings):
    openapi_url: str = "/openapi.json"


settings = Settings()

app = FastAPI(openapi_url=settings.openapi_url)


@app.get("/")
def root():
    return {"message": "Hello World"}
```


Then you could call it with:

```console
$ OPENAPI_URL= uvicorn main:app
```

**Pull request closed without merging**